src/window.rs

```diff
@@ -560,18 +560,80 @@ pub struct Window<T: Layout> {
 #[derive(Debug, Copy, Clone)]
 pub struct AnimationState { }
 
+pub struct ScrollStates(pub FastHashMap<ExternalScrollId, ScrollState>);
+
+impl ScrollStates {
+    pub fn new() -> ScrollStates {
+        ScrollStates(FastHashMap::default())
+    }
+
+    /// NOTE: This has to be a getter, because we need to update
+    #[must_use]
+    pub(crate) fn get_scroll_amount(&mut self, scroll_id: &ExternalScrollId) -> Option<(f32, f32)> {
+        let entry = self.0.get_mut(&scroll_id)?;
+        Some(entry.get())
+    }
+
+    /// Updating the scroll amount does not update the `entry.used_this_frame`,
+    /// since that is only relevant when we are actually querying the renderer.
+    pub(crate) fn scroll_node(&mut self, scroll_id: &ExternalScrollId, scroll_by_x: f32, scroll_by_y: f32) {
+        if let Some(entry) = self.0.get_mut(scroll_id) {
+            entry.add(scroll_by_x, scroll_by_y);
+        }
+    }
+
+    pub(crate) fn ensure_initialized_scroll_state(&mut self, scroll_id: ExternalScrollId, overflow_x: f32, overflow_y: f32) {
+        if !self.0.contains_key(&scroll_id) {
```

Comment:
Use the `.entry` API instead, this way you can avoid the double lookup (both `.contains_key` and `.insert` have to lookup the key in the HashMap):

```rust
self.0.entry(&scroll_id).or_insert_with(|| ScrollState::new(overflow_x, overflow_y));
```

Reply:
Oh wow, this is cool, thanks!
FYI: HashMap::entry(key: TKey) takes an owned key, not a borrowed one ;)